velox/functions/prestosql/ArrayShuffle.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <numeric>
+#include <random>
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/expression/VectorReaders.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+// See documentation at
+// https://prestodb.io/docs/current/functions/array.html#shuffle
+template <TypeKind T>
+class ArrayShuffleFunction : public exec::VectorFunction {
+ public:
+  ArrayShuffleFunction() {}
+
+  bool isDeterministic() const override {
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /*outputType*/,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    auto& arg = args[0];
+    VectorPtr localResult;
+
+    // Input can be constant, flat or dictionary.
+    if (arg->isConstantEncoding()) {
+      auto* constantArray = arg->as<ConstantVector<ComplexType>>();
+      const auto& flatArray = constantArray->valueVector();
+      const auto flatIndex = constantArray->index();
+
+      exec::LocalSelectivityVector singleRow(context, flatIndex + 1);
+      singleRow->setValid(flatIndex, true);
+      singleRow->updateBounds();
+
+      localResult = applyFlat(
+          *singleRow.get(), flatArray.get()->as<ArrayVector>(), context);
+      localResult =
+          BaseVector::wrapInConstant(rows.size(), flatIndex, localResult);
+    } else if (arg->isFlatEncoding()) {
+      localResult = applyFlat(rows, arg->as<ArrayVector>(), context);
+    } else { // dictionary encoding
+      exec::LocalDecodedVector decodedVectorHolder(context, *arg, rows);
+      localResult = applyFlat(
+          rows, decodedVectorHolder->base()->as<ArrayVector>(), context);
```

Comment:
This doesn't look right as 'rows' refer to top-level vector, but applyFlat function receives the base vector. For example, top-level vector may have rows 0, 1, ....10, but the the base vector may have only 2 rows: 0, 1. Accesssing rows 2,...10 will cause an error.

Reply:
Thx for pointing out. I should merge the flat/dict paths, and use `LocalDecodedVector` to make the encoding handling transparent.